Firestore/core/src/firebase/firestore/remote/datastore.mm

```diff
@@ -99,16 +123,115 @@
-  return std::make_shared<WriteStream>(worker_queue_, credentials_, serializer_,
+  return std::make_shared<WriteStream>(worker_queue_, credentials_,
+                                       serializer_bridge_.GetSerializer(),
                                        &grpc_connection_, delegate);
 }
 
+void Datastore::LookupDocuments(
+    const std::vector<DocumentKey> &keys,
+    FSTVoidMaybeDocumentArrayErrorBlock completion) {
+  grpc::ByteBuffer message = serializer_bridge_.ToByteBuffer(
+      serializer_bridge_.CreateLookupRequest(keys));
+
+  WithToken(
+      [this, message, completion](const Token &token) {
+        LookupDocumentsWithToken(token, message, completion);
+      },
+      [completion](const Status &status) {
+        completion(nil, util::MakeNSError(status));
+      });
+}
+
+void Datastore::LookupDocumentsWithToken(
+    const Token &token, const grpc::ByteBuffer &message,
+    FSTVoidMaybeDocumentArrayErrorBlock completion
+    ) {
+  lookup_calls_.push_back(grpc_connection_.CreateStreamingReader(
+      kRpcNameLookup, token, std::move(message)));
+  GrpcStreamingReader *call = lookup_calls_.back().get();
+
+  call->Start([this, call, completion](const Status &status,
+                           const std::vector<grpc::ByteBuffer> &responses) {
+    OnLookupDocumentsResponse(call, status, responses, completion);
+    RemoveGrpcCall(call);
+  });
+}
+
+void Datastore::OnLookupDocumentsResponse(
+    GrpcStreamingReader *call,
+    const Status &status,
+    const std::vector<grpc::ByteBuffer> &responses,
+    FSTVoidMaybeDocumentArrayErrorBlock completion
+    ) {
+  LogGrpcCallFinished("BatchGetDocuments", call, status);
+  HandleCallStatus(status);
+
+  if (!status.ok()) {
+    completion(nil, util::MakeNSError(status));
+    return;
+  }
+
+  Status parse_status;
+  NSArray<FSTMaybeDocument *> *docs =
+      serializer_bridge_.MergeLookupResponses(responses, &parse_status);
+  if (parse_status.ok()) {
+    completion(docs, nil);
+  } else {
+    completion(nil, util::MakeNSError(parse_status));
+  }
+}
+
+void Datastore::WithToken(const OnToken &on_token, const OnError &on_error) {
+  // Auth may outlive Firestore
+  std::weak_ptr<Datastore> weak_this{shared_from_this()};
+
+  credentials_->GetToken(
+      [weak_this, on_token, on_error](StatusOr<Token> maybe_token) {
+        auto strong_this = weak_this.lock();
+        if (!strong_this) {
+          return;
+        }
+
+        strong_this->worker_queue_->EnqueueRelaxed(
+            [weak_this, maybe_token, on_token, on_error] {
+              auto strong_this = weak_this.lock();
+              if (!strong_this) {
+                return;
+              }
+
+              if (maybe_token.ok()) {
+                on_token(maybe_token.ValueOrDie());
+              } else {
+                on_error(maybe_token.status());
+              }
+            });
+      });
+}
+
+void Datastore::HandleCallStatus(const Status &status) {
```

Comment:
This function will be reused in the other `Datastore` RPC (`CommitMutations`), so I'd prefer not to inline it. While I thought about naming it `MaybeInvalidateToken`, I chose the more vague name sort of deliberately, because I want to keep any potential status handling in this function. While currently it's only the token invalidation, Datastore could in theory react to other statuses as well. I was also trying to keep this consistent with [`Stream`](https://github.com/firebase/firebase-ios-sdk/blob/master/Firestore/core/src/firebase/firestore/remote/stream.mm#L280-L291).